modules/internal/ExternalArray.chpl

```diff
@@ -55,7 +55,8 @@ module ExternalArray {
         halt("there should only be one set of indices, not multiple dimensions");
       }
       var r = inds(1);
-      if (r.low != 0) {
+      // Allow empty arrays (which have the domain 1..0)
+      if (r.low != 0 && !(r.low == 1 && r.high == 0)) {
         halt("external arrays always have a lower bound of 0");
```

Comment:
Should it say non-empty external arrays have a lower bound of 0?
Maybe do one more check that we're not relying on the lower bound.

Reply:
I'm a little confused, can you clarify what you're thinking here?

Reply:
1..0 bounds would mean the lower bound could be 0, couldn't it?

Reply:
Yes.  I think the error message change makes sense.  Should have clarified, I'm mostly confused about the "extra check" portion of that comment

Reply:
> Maybe do one more check that we're not relying on the lower bound.

Is there code in this module that won't work if r.low == 1 and the array is actually nil?

Reply:
I don't think so.  The rest of the implementation seems to only check on the size of the domain, which is stored separately instead of as a range.  This code doesn't pass the range on, just its size.

It looks like ExternDom's domRange method returns 0..#0/1..0 (which dsiGetIndices and dsiDim(s) relies on) but seems to be used in loops where an empty range would just mean the loop got skipped anyways

Reply:
That said, I don't fully understand the purpose of domRange/dsiGetIndices/dsiDim/dsiDims, so it's possible there is something I'm missing

Reply:
Circling back now that I am on top of things again, do you think there is an additional action item for this PR or is it good to go?  (I know there's a green approval check mark but this discussion made me uncertain)

Reply:
I'm OK with it, just wanted you to think about it, which you did